File: assets/scripts/Python/roc_utils.py
# ================================================================
# File: roc_utils.py
# Author: Xi Hu
# Description:
#   Utility functions for ROC analysis and visualization:
#   (1) Compute ROC AUC and 95% Confidence Interval (bootstrapped)
#   (2) Plot ROC curves for multiple datasets (faceted)
#   (3) Plot multiple ROC curves in one overlay panel
# Dependencies: numpy, pandas, matplotlib, scikit-learn
# ================================================================

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from matplotlib.ticker import AutoMinorLocator, FixedLocator
from sklearn.metrics import roc_curve, roc_auc_score


# ================================================================
# Function: roc_auc_ci
# Purpose : Compute ROC AUC and its 95% Confidence Interval
# ================================================================
def roc_auc_ci(y_true=None, y_score=None, help=False, n_bootstrap=2000, random_state=42):
    """
    Compute the ROC AUC value and its 95% confidence interval using bootstrapping.

    Args:
        y_true (array-like): True binary labels (0 or 1).
        y_score (array-like): Predicted continuous scores or probabilities.
        help (bool): If True, print this docstring and return None.
        n_bootstrap (int): Number of bootstrap resamples for CI estimation.
        random_state (int): Random seed for reproducibility.

    Returns:
        dict: {
            "AUC": float,           # Calculated AUC value
            "CI": (lower, mean, upper)
        }

    Example:
        >>> result = roc_auc_ci([0, 1, 1, 0], [0.2, 0.8, 0.7, 0.3])
        >>> print(result["AUC"], result["CI"])
    """
    if help:
        print(roc_auc_ci.__doc__)
        return None

    y_true = np.array(y_true)
    y_score = np.array(y_score)
    auc_val = roc_auc_score(y_true, y_score)

    rng = np.random.RandomState(random_state)
    boot_scores = []

    for _ in range(n_bootstrap):
        indices = rng.randint(0, len(y_score), len(y_score))
        if len(np.unique(y_true[indices])) < 2:
            continue
        boot_auc = roc_auc_score(y_true[indices], y_score[indices])
        boot_scores.append(boot_auc)

    sorted_scores = np.sort(boot_scores)
    ci_lower = np.percentile(sorted_scores, 2.5)
    ci_upper = np.percentile(sorted_scores, 97.5)

    return {"AUC": auc_val, "CI": (ci_lower, auc_val, ci_upper)}


# ================================================================
# Function: plot_ROC_spine
# Purpose : Apply consistent axis/tick/label style for ROC plots
# ================================================================
def plot_ROC_spine(ax, base_font=22, major_len=8, minor_len=5, minor_step=0.05, fixedlocator=True):
    """
    Apply a consistent, publication-style format for ROC plots.

    Features:
        - Label axes with "Sensitivity (%)" and "Specificity (%)"
        - Show ticks from 100→0 (x-axis) and 0→100 (y-axis)
        - Draw outward major/minor ticks
        - Add diagonal reference line

    Args:
        ax (matplotlib.axes): Target axes object.
        base_font (int): Base font size for labels/ticks.
        major_len (int): Major tick length in points.
        minor_len (int): Minor tick length in points.
        minor_step (float): Step for minor ticks if fixed locator is used.
        fixedlocator (bool): If True, use fixed 0–1 minor ticks.
    """
    # Axis labels
    ax.set_ylabel('Sensitivity (%)', fontsize=base_font)
    ax.set_xlabel('Specificity (%)', fontsize=base_font)

    # Major tick labels
    ax.yaxis.set_tick_params(labelsize=base_font-7)
    ax.xaxis.set_tick_params(labelsize=base_font-7)

    # Major ticks: outward style
    ax.tick_params(direction='out', length=major_len, width=1,
                    colors='k', grid_color='none', grid_alpha=0.5)

    # Define major tick positions
    major = np.arange(0, 1.1, 0.2)
    ax.set_xlim([-0.05, 1.05])
    ax.set_ylim([-0.05, 1.05])
    ax.set_xticks(major)
    ax.set_xticklabels([f'{(1-x)*100:.0f}' for x in major])  # 100→0
    ax.set_yticks(major)
    ax.set_yticklabels([f'{y*100:.0f}' for y in major])      # 0→100

    # Optional: fixed minor ticks within [0, 1]
    if fixedlocator:
        minor_ticks = np.arange(0.0, 1.0 + 1e-9, minor_step)
        ax.xaxis.set_minor_locator(FixedLocator(minor_ticks))
        ax.yaxis.set_minor_locator(FixedLocator(minor_ticks))
        ax.tick_params(which='minor', direction='out', length=minor_len, width=1, colors='k')

    # Add diagonal reference line
    ax.plot([-0.05, 1.05], [-0.05, 1.05], color='grey', lw=1.2, linestyle='--')
    return ax


# ================================================================
# Function: plot_roc_overlay_pub
# Purpose : Overlay multiple ROC curves in one figure
# ================================================================
def plot_roc_overlay_pub(
    infer_df,
    datasets,
    truth_col="true",
    score_col="score",
    colors=None,
    title_main="ROC Curve Comparison",
    base_font=22,
    title_font=None,
    line_width=2.5,
    alpha=1.0,
    figsize=(8,8),
    legend_title="AUC",
    legend_font_delta=-7,
    legend_loc=(0.78, 0.20),
    fixedlocator=True,
):
    """
    Plot multiple ROC curves in one overlay figure with consistent styling.

    Args:
        infer_df (pd.DataFrame): Data containing 'Dataset', true, and score columns.
        datasets (list): Names of datasets to include in the overlay.
        truth_col (str): Column name for true labels.
        score_col (str): Column name for predicted scores.
        colors (dict or None): Optional mapping {dataset: color}.
        title_main (str): Main title for the plot.
        base_font (int): Base font size for axes/labels.
        line_width (float): ROC line width.
        alpha (float): Line transparency.
        figsize (tuple): Figure size (width, height).
        legend_title (str): Title for legend block.
        legend_font_delta (int): Relative font size for legend text.
        legend_loc (tuple): Legend position anchor (x, y) in axis coordinates.
        fixedlocator (bool): If True, use fixed minor ticks between 0–1.

    Returns:
        (fig, ax): Matplotlib figure and axis objects.
    """
    fig, ax = plt.subplots(figsize=figsize)

    # Plot each dataset
    for ds in datasets:
        sub = infer_df[infer_df["Dataset"] == ds]
        y_true = sub[truth_col].to_numpy()
        y_score = sub[score_col].to_numpy()
        fpr, tpr, _ = roc_curve(y_true, y_score)
        auc_val = roc_auc_score(y_true, y_score)

        color = colors.get(ds, None) if isinstance(colors, dict) else None
        ax.plot(
            fpr, tpr,
            color=color, lw=line_width, alpha=alpha,
            label=f'{ds} (AUC = {auc_val:.2f})'
        )

    # Legend block
    leg = ax.legend(
        frameon=False,
        fontsize=max(6, base_font + legend_font_delta),
        title=legend_title,
        title_fontsize=max(6, base_font + legend_font_delta + 3),
        loc='center',
        bbox_to_anchor=legend_loc
    )
    try:
        leg._legend_box.align = "right"
    except Exception:
        pass

    # Apply unified ROC axis style
    if not fixedlocator:
        ax.yaxis.set_minor_locator(AutoMinorLocator())
        ax.xaxis.set_minor_locator(AutoMinorLocator())
    plot_ROC_spine(ax, base_font=base_font, fixedlocator=fixedlocator)

    # Reinforce tick visuals
    ax.tick_params(direction='out', length=10, width=1, colors='k', grid_color='grey', grid_alpha=1)
    ax.tick_params(which='minor', length=5, width=1, color='k')

    ax.set_title(title_main, fontsize=title_font or base_font, weight='bold', pad=8)
    fig.canvas.draw_idle()
    return fig, ax


# ================================================================
# Function: plot_roc_panels_pub
# Purpose : Plot multiple ROC curves in separate subplots (faceted layout)
# ================================================================
def plot_roc_panels_pub(
    infer_df,
    datasets=None,
    truth_col="true",
    score_col="score",
    colors=None,
    title_main="ROC Curve Panels",
    base_font=20,
    title_font=None,
    line_width=2.0,
    alpha=1.0,
    ncol=3,
    panel_size=(5, 5),
    legend_title="AUC",
    legend_font_delta=-6,
    fixedlocator=True,
):
    """
    Plot multiple ROC curves in a faceted (multi-panel) layout.

    Each dataset is plotted in its own subplot with consistent
    styling (axes, ticks, legend placement, etc.), suitable for
    figure panels in publications.

    Args:
        infer_df (pd.DataFrame): Data containing [Dataset, truth, score].
        datasets (list or None): Dataset names to plot (default: all).
        truth_col (str): Column name for true labels.
        score_col (str): Column name for predicted scores.
        colors (dict): Optional {dataset: color} mapping.
        title_main (str): Global figure title.
        base_font (int): Font size for labels and titles.
        line_width (float): Width of ROC line.
        alpha (float): Transparency for ROC line.
        ncol (int): Number of subplot columns.
        panel_size (tuple): Each subplot size in inches (width, height).
        legend_title (str): Title for the legend.
        legend_font_delta (int): Relative font size adjustment for legend.
        fixedlocator (bool): Use fixed 0–1 minor tick range.

    Returns:
        (fig, axes): Matplotlib figure and axes array.
    """
    if datasets is None:
        datasets = infer_df["Dataset"].unique().tolist()

    n_datasets = len(datasets)
    nrow = int(np.ceil(n_datasets / ncol))

    fig, axes = plt.subplots(
        nrow, ncol,
        figsize=(panel_size[0] * ncol, panel_size[1] * nrow),
        squeeze=False
    )

    for idx, ds in enumerate(datasets):
        r, c = divmod(idx, ncol)
        ax = axes[r, c]

        sub = infer_df[infer_df["Dataset"] == ds]
        y_true = sub[truth_col].to_numpy()
        y_score = sub[score_col].to_numpy()
        fpr, tpr, _ = roc_curve(y_true, y_score)
        auc_val = roc_auc_score(y_true, y_score)

        color = colors.get(ds, None) if isinstance(colors, dict) else None
        ax.plot(
            fpr, tpr,
            color=color,
            lw=line_width,
            alpha=alpha,
            label=f"AUC = {auc_val:.2f}"
        )

        # Local legend
        leg = ax.legend(
            frameon=False,
            fontsize=max(6, base_font + legend_font_delta),
            title=legend_title,
            title_fontsize=max(6, base_font + legend_font_delta + 2),
            loc='lower right'
        )
        try:
            leg._legend_box.align = "right"
        except Exception:
            pass

        # Apply consistent ROC styling
        plot_ROC_spine(ax, base_font=base_font, fixedlocator=fixedlocator)

        # Reinforce tick visuals
        ax.tick_params(direction='out', length=10, width=1, colors='k', grid_color='grey', grid_alpha=1)
        ax.tick_params(which='minor', length=5, width=1, color='k')

        ax.set_title(ds, fontsize=title_font or base_font, weight='bold', pad=5)

    # Remove empty subplots if any
    total_axes = nrow * ncol
    for j in range(n_datasets, total_axes):
        fig.delaxes(axes.flatten()[j])

    # Global layout & title
    fig.suptitle(title_main, fontsize=title_font or base_font + 4, weight='bold', y=0.98)
    fig.tight_layout(rect=[0.05, 0.05, .95, 0.95])
    fig.canvas.draw_idle()
    return fig, axes


# ================================================================
# Example Section (Demo)
# ================================================================
if __name__ == "__main__":
    np.random.seed(2025)
    n = 300

    # --- Generate demo datasets with varying separability ---
    datasets = {
        "Linear_High": pd.DataFrame({
            "Dataset": "Linear_High",
            "true": np.repeat([0, 1], n // 2),
            "score": np.concatenate([
                np.random.normal(0, 1, n // 2),
                np.random.normal(3, 1, n // 2)
            ])
        }),
        "Linear_Mid": pd.DataFrame({
            "Dataset": "Linear_Mid",
            "true": np.repeat([0, 1], n // 2),
            "score": np.concatenate([
                np.random.normal(0, 1, n // 2),
                np.random.normal(1.5, 1, n // 2)
            ])
        }),
        "Linear_Low": pd.DataFrame({
            "Dataset": "Linear_Low",
            "true": np.repeat([0, 1], n // 2),
            "score": np.concatenate([
                np.random.normal(0, 1, n // 2),
                np.random.normal(0.5, 1, n // 2)
            ])
        })
    }

    demo = pd.concat(datasets.values(), ignore_index=True)

    custom_colors = {
        "Linear_High": "#E64B35",
        "Linear_Mid": "#4DD576",
        "Linear_Low": "#1C97CC"
    }

    # --- Overlay ROC curves (single panel) ---
    fig, ax = plot_roc_overlay_pub(
        infer_df=demo,
        datasets=["Linear_High", "Linear_Mid", "Linear_Low"],
        colors=custom_colors,
        title_main="ROC Curve Demo (High / Mid / Low Separability)",
        base_font=22,
        line_width=2.0,
        figsize=(8, 8),
        legend_loc=(0.7, 0.20)
    )
    plt.show()

    # --- Faceted ROC panels (multi-panel layout) ---
    fig_panels, axes = plot_roc_panels_pub(
        infer_df=demo,
        datasets=["Linear_High", "Linear_Mid", "Linear_Low"],
        colors=custom_colors,
        title_main="ROC Curve Panels (High / Mid / Low)",
        base_font=18,
        ncol=2,
        panel_size=(5, 5)
    )
    plt.show()
    

File: assets/scripts/R/roc_utils.R
# ================================================================
# File: roc_utils.R
# Author: Xi HU
# Description:
#   Utility functions for ROC analysis:
#   (1) Compute ROC AUC and 95% Confidence Interval.
#   (2) Plot ROC curves for multiple datasets in one panel.
#   (3) Plot multiple ROC curves in a single panel (with AUC legend).
# Dependencies: pROC, ggplot2, dplyr, tidyr, scales
# ================================================================

library(pROC)
library(ggplot2)
library(dplyr)
library(scales)

# ================================================================
# Function: roc_auc_ci
# Purpose : Compute ROC AUC and its 95% Confidence Interval
# ================================================================
roc_auc_ci <- function(y_true=NULL, y_score=NULL, help=FALSE) {
	if (help) {
		cat("
	Compute ROC AUC and its 95% Confidence Interval
	------------------------------------------------
	
	Description:
        Calculates the area under the ROC curve (AUC) and its 95% confidence interval
        using the `pROC` package.
	
	Arguments:
        y_true  : Numeric vector of true binary labels (0 or 1).
        y_score : Numeric vector of predicted scores or probabilities.
        help    : Logical flag. If TRUE, prints this help text and exits.
	
	Returns:
        A list containing:
        $AUC : numeric, area under the ROC curve.
        $CI  : numeric vector (lower, mean, upper) for the 95% CI.
	
	Example:
        result <- roc_auc_ci(c(0,1,1,0), c(0.2,0.8,0.7,0.3))
        result$AUC
        result$CI
	
	------------------------------------------------
	")
	return(invisible(NULL))
	}
	
	library(pROC)
	
	roc_obj <- roc(y_true, y_score, quiet = TRUE)
	auc_val <- auc(roc_obj)
	ci_val <- ci.auc(roc_obj)
	return(list(AUC = auc_val, CI = ci_val))
}

# ================================================================
# Function: plot_roc_panels
# Purpose : Plot ROC Curves for Multiple Datasets
# ================================================================
plot_roc_panels <- function(
	infer_df,
	datasets = c("Cohort 01", "Cohort 02"),
	truth_col = "true",
	score_col = "score",
	colors = c("#F8766D"),
	line_width = 1.2,
	line_alpha = 0.9,
	title_main = "ROC Curve Analysis",
	x_label = "Specificity",
	y_label = "Sensitivity",
	show_auc_label = TRUE,
	auc_pos = c(0.35, 0.18),
	auc_digits = 3,
	ncol = 2,
	base_font = 16,
	help=FALSE
) {
	if (help) {
	cat("
		Plot ROC Curves for Multiple Datasets
		------------------------------------------------
		
		Description:
            Generates faceted ROC plots for one or more datasets using ggplot2.
            It computes ROC curves, AUC, and 95% CI for each dataset, and can
            annotate AUC values directly on the plot.
        
        Arguments:
            infer_df       : Data frame containing true labels, scores, and dataset IDs.
            datasets       : Character vector specifying which dataset names to plot.
            truth_col      : Column name for true binary labels (0/1).
            score_col      : Column name for predicted scores or probabilities.
            colors         : Vector of colors for ROC curves.
            line_width     : Numeric, width of the ROC curve line.
            line_alpha     : Numeric between 0–1, transparency level.
            title_main     : Main title of the plot.
            x_label        : Label for x-axis (default = 'Specificity').
            y_label        : Label for y-axis (default = 'Sensitivity').
            show_auc_label : Logical, whether to display AUC text on plot.
            auc_pos        : Numeric vector (x, y) for AUC label position.
            auc_digits     : Integer, number of decimals for AUC and CI.
            ncol           : Number of facet columns.
            base_font      : Base font size for ggplot theme.
            help           : Logical flag. If TRUE, prints this help text and exits.
		
		Returns:
            A ggplot object showing ROC curves (one facet per dataset).
		
		Example:
            data(pROC::aSAH)
            demo <- aSAH %>%
            dplyr::mutate(true = ifelse(outcome == 'Poor', 1, 0)) %>%
            dplyr::select(true, s100b, ndka)
		
            demo_long <- demo %>%
            tidyr::pivot_longer(cols = c(s100b, ndka),
                                names_to = 'Dataset',
                                values_to = 'score')
		
        demo_long$Dataset <- factor(demo_long$Dataset, levels = c('s100b', 'ndka'))
		
        p <- plot_roc_panels(
                infer_df   = demo_long,
                datasets   = c('s100b','ndka'),
                truth_col  = 'true',
                score_col  = 'score',
                title_main = 'ROC Curve Demo (aSAH)',
                base_font  = 15
            )
        print(p)
		------------------------------------------------
		")
	return(invisible(NULL))
	}
	
	library(ggplot2)
	library(dplyr)
	library(pROC)
	
	roc_list <- lapply(datasets, function(ds) {
	idf <- infer_df %>% filter(.data$Dataset == ds)
	roc_obj <- pROC::roc(
		response = idf[[truth_col]],
		predictor = idf[[score_col]],
		quiet = TRUE
	)
	data.frame(
        Dataset = ds,
        Specificity = as.numeric(roc_obj$specificities),
        Sensitivity = as.numeric(roc_obj$sensitivities),
        AUC = as.numeric(pROC::auc(roc_obj)),
        CI_low = as.numeric(pROC::ci.auc(roc_obj))[1],
        CI_high = as.numeric(pROC::ci.auc(roc_obj))[3]
	)
	})
	
	roc_df <- do.call(rbind, roc_list)
	
	ann_df <- roc_df %>%
	group_by(Dataset) %>%
	summarise(AUC = first(AUC), CI_low = first(CI_low),
			CI_high = first(CI_high), .groups = "drop") %>%
	mutate(
	x = auc_pos[1],
	y = auc_pos[2],
	label = sprintf("AUC = %.3f\n95%% CI: [%.3f, %.3f]",
                    round(AUC, auc_digits),
                    round(CI_low, auc_digits),
                    round(CI_high, auc_digits))
	)
	
	p <- ggplot(roc_df, aes(x = Specificity, y = Sensitivity)) +
	geom_path(
        color = colors[1],
        linewidth = line_width,
        alpha = line_alpha,
        lineend = "round"
	) +
	annotate("segment",
                x = 1, y = 0, xend = 0, yend = 1,
                linetype = "dashed", color = "grey70") +
	scale_x_reverse(limits = c(1, 0), breaks = seq(1, 0, by = -0.2)) +
	scale_y_continuous(limits = c(0, 1), breaks = seq(0, 1, by = 0.2)) +
	coord_fixed(ratio = 1) +
	facet_wrap(~ Dataset, ncol = ncol, scales = "fixed") +
	labs(title = title_main, x = x_label, y = y_label) +
	theme_bw(base_size = base_font) +
	theme(
        plot.title = element_text(hjust = 0.5, face = "bold", size = base_font + 2),
        panel.grid = element_blank(),
        axis.title = element_text(size = base_font),
        axis.text = element_text(size = base_font - 2),
        strip.background = element_rect(fill = "grey85", color = NA),
        strip.text = element_text(size = base_font - 2)
	)
	
	if (show_auc_label) {
	p <- p + geom_text(
        data = ann_df,
        aes(x = x, y = y, label = label),
        inherit.aes = FALSE,
        size = base_font / 3.5
	)
	}
	
	return(p)
}

# ================================================================
# Function: plot_roc_overlay
# Purpose : Plot multiple ROC curves in a single panel (with AUC legend)
# ================================================================
plot_roc_overlay <- function(
	infer_df = NULL,
	datasets = NULL,
	truth_col = "true",
	score_col = "score",
	colors = NULL,
	line_width = 1.2,
	line_alpha = 0.9,
	title_main = "ROC Curve Comparison",
	x_label = "Specificity",
	y_label = "Sensitivity",
	base_font = 16,
	legend_position='right',
	help = FALSE
	) {
	if (help) {
	cat("
		Plot Multiple ROC Curves in One Panel
		------------------------------------------------
		
		Description:
            Plots ROC curves from multiple datasets (or groups) together
            in a single panel using ggplot2. Each curve is labeled in the
            legend as 'Dataset (AUC = 0.xx)'.
        
        Arguments:
            infer_df    : Data frame containing true labels, scores, and dataset IDs.
            datasets    : Character vector specifying dataset names to include.
                        If NULL, all unique values of 'Dataset' column are used.
            truth_col   : Column name for true binary labels (0/1).
            score_col   : Column name for predicted scores or probabilities.
            colors      : Vector of colors for each dataset. If NULL, ggplot default palette used.
            line_width  : Numeric, width of ROC lines.
            line_alpha  : Numeric between 0–1, transparency of ROC lines.
            title_main  : Main title of the plot.
            x_label, y_label : Axis labels.
            base_font   : Base font size for ggplot theme.
            help        : Logical flag. If TRUE, prints this help text and exits.
		
		Returns:
            A ggplot object showing all ROC curves overlaid in one figure.
        
        Example:
            data(pROC::aSAH)
            demo <- aSAH %>%
            dplyr::mutate(true = ifelse(outcome == 'Poor', 1, 0)) %>%
            dplyr::select(true, s100b, ndka)
        
            demo_long <- demo %>%
            tidyr::pivot_longer(
                cols = c(s100b, ndka),
                names_to = 'Dataset',
                values_to = 'score'
            )
        
            plot_roc_overlay(
            infer_df   = demo_long,
            datasets   = c('s100b', 'ndka'),
            truth_col  = 'true',
            score_col  = 'score',
            title_main = 'ROC Comparison: s100b vs ndka'
            )
		------------------------------------------------
		")
	return(invisible(NULL))
	}

	library(pROC)
	library(ggplot2)
	library(dplyr)
	library(scales)

	# Automatically detect datasets if not specified
	if (is.null(datasets)) {
	datasets <- unique(infer_df$Dataset)
	}
	
	# Compute ROC for each dataset
	roc_list <- lapply(datasets, function(ds) {
	idf <- infer_df %>% dplyr::filter(.data$Dataset == ds)
	roc_obj <- pROC::roc(idf[[truth_col]], idf[[score_col]], quiet = TRUE)
	data.frame(
        Dataset = ds,
        Specificity = as.numeric(roc_obj$specificities),
        Sensitivity = as.numeric(roc_obj$sensitivities),
        AUC = as.numeric(pROC::auc(roc_obj))
	)
	})
	
	# Combine results
    roc_df <- do.call(rbind, roc_list) %>%
        dplyr::mutate(label = sprintf("%s (AUC = %.2f)", Dataset, AUC)) %>%
        dplyr::mutate(label = factor(label, 
                                levels = unique(label[order(-AUC)])))
	
	# Color palette
	if (is.null(colors)) {
	colors <- scales::hue_pal()(length(datasets))
	}
	
	# Plot all ROC curves together
	p <- ggplot(roc_df, aes(x = Specificity, y = Sensitivity, color = label)) +
	geom_path(linewidth = line_width, alpha = line_alpha, lineend = "round") +
	annotate("segment", x = 1, y = 0, xend = 0, yend = 1,
                linetype = "dashed", color = "grey60") +
	scale_x_reverse(limits = c(1, 0), breaks = seq(1, 0, by = -0.2)) +
	scale_y_continuous(limits = c(0, 1), breaks = seq(0, 1, by = 0.2)) +
	coord_fixed(ratio = 1) +
	labs(
        title = title_main,
        x = x_label,
        y = y_label,
        color = "Groups"
	) +
	theme_bw(base_size = base_font) +
	theme(
		plot.title = element_text(hjust = 0.5, face = "bold", size = base_font + 2),
		panel.grid = element_blank(),
		axis.title = element_text(size = base_font),
		axis.text = element_text(size = base_font - 2),
		legend.title = element_text(size = base_font - 1, face = "bold"),
		legend.text = element_text(size = base_font - 2),
		legend.position=legend_position,
	)
	
	return(p)
}
